### PR TITLE
overwrite segments for removed code

### PR DESCRIPTION
**Pull request by aleclarson**
Removed ranges create two source map segments that refer to the same generated column. I consider this bloat if there isn't a good reason behind it. Generally, there shouldn't be segments for removed code (not including the segment for the characters after, whose source column is the removed range's end column).

What's your take on this, @mourner and/or @Rich-Harris?

Figured I should get feedback before writing any tests.

### CODE REVIEW

**Comment by mourner** [comment]
@guybedford @lukastaegert what do you think? I'm not sure I fully understand the implications of this change for downstream users, so I'd refer to others for review.

File: src/utils/Mappings.js [hunk 1]
```diff
@@ -27,12 +27,20 @@ export default class Mappings {
 		let originalCharIndex = chunk.start;
 		let first = true;
 
+		let len = this.rawSegments.length;
```
Comment (guybedford) [inline src/utils/Mappings.js]: Can we rather call this something more explanatory, perhaps:

```js
let checkDuplicateSegmentIndex = this.rawSegments.length - 1;
```

File: src/utils/Mappings.js [hunk 2]
```diff
@@ -27,12 +27,20 @@ export default class Mappings {
...
 		while (originalCharIndex < chunk.end) {
 			if (this.hires || first || sourcemapLocations[originalCharIndex]) {
-				this.rawSegments.push([this.generatedCodeColumn, sourceIndex, loc.line, loc.column]);
+				if (len !== 0) {
```
Comment (guybedford) [inline src/utils/Mappings.js]: This would then become `if (checkDuplpicateSegmentIndex !== -1)`

File: src/utils/Mappings.js [hunk 3]
```diff
@@ -27,12 +27,20 @@ export default class Mappings {
...
+					const segment = this.rawSegments[len - 1];
+					if (segment[0] === this.generatedCodeColumn) {
+						len -= 1; // overwrite segments for removed code
```
Comment (guybedford) [inline src/utils/Mappings.js]: These changes to `len` can be unnecessary if you just modify the last segment directly:

```js
this.rawSegments[checkDuplicateSegmentIndex] = [...];
```

File: src/utils/Mappings.js [hunk 4]
```diff
@@ -27,12 +27,20 @@ export default class Mappings {
...
+					}
+				}
+				this.rawSegments[len++] = [this.generatedCodeColumn, sourceIndex, loc.line, loc.column];
```
Comment (guybedford) [inline src/utils/Mappings.js]: The default path should be able to then revert to a `.push` as before.

File: src/utils/Mappings.js [hunk 5]
```diff
@@ -27,12 +27,20 @@ export default class Mappings {
...
 			}
 
 			if (original[originalCharIndex] === '\n') {
+				len = 0;
```
Comment (guybedford) [inline src/utils/Mappings.js]: And this becomes `checkDuplicateSegmentIndex = -1` making it clearer this is a switch on the loop.

**Review by guybedford — CHANGES_REQUESTED** [review]
Thanks for much @aleclarson for this PR.

This looks great in principle to me, but the test needs to be updated, and it would ideally come with another test that considers if there are any pathological edge cases of this behaviour - let me know if you can think of any interesting things that may come up here.

**Comment by antfu** [comment]
Cleanup due to staleness. Thanks for the contribution
